azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -715,3 +718,460 @@ def test_ignores_unknown_request_id(self, stage, op, iot_response, unhandled_err
         stage.next._send_event_up(iot_response)
         assert op.callback.call_count == 0
         assert unhandled_error_handler.call_count == 0
+
+
+"""
+A note on terms in the TimeoutStage tests:
+    No-timeout ops are ops that don't need a timeout check
+    Yes-timeout ops are ops that do need a timeout check
+"""
+timeout_intervals = {
+    pipeline_ops_mqtt.MQTTSubscribeOperation: 10,
+    pipeline_ops_mqtt.MQTTUnsubscribeOperation: 10,
+}
+yes_timeout_ops = list(timeout_intervals.keys())
+no_timeout_ops = all_except(all_common_ops, yes_timeout_ops)
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    cls=pipeline_stages_base.TimeoutStage,
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=yes_timeout_ops,
+    all_events=all_common_events,
+    handled_events=[],
+    extra_initializer_defaults={"timeout_intervals": timeout_intervals},
+)
+
+
+@pytest.fixture()
+def mock_timer(mocker):
+    return mocker.patch(
+        "azure.iot.device.common.pipeline.pipeline_stages_base.Timer", autospec=True
+    )
+
+
+@pytest.mark.describe("TimeoutStage - run_op()")
+class TestTimeoutStageRunOp(StageTestBase):
+    @pytest.fixture(params=yes_timeout_ops)
+    def yes_timeout_op(self, request, mocker):
+        op = make_mock_op_or_event(request.param)
+        op.callback = mocker.MagicMock()
+        return op
+
+    @pytest.fixture(params=no_timeout_ops)
+    def no_timeout_op(self, request, mocker):
+        op = make_mock_op_or_event(request.param)
+        op.callback = mocker.MagicMock()
+        return op
+
+    @pytest.fixture
+    def stage(self):
+        return pipeline_stages_base.TimeoutStage()
+
+    @pytest.mark.it("Sends ops that don't need a timer to the next stage")
+    def test_sends_no_timer_op_down(self, stage, mock_timer, no_timeout_op):
+        stage.run_op(no_timeout_op)
+        assert stage.next.run_op.call_count == 1
+        assert stage.next.run_op.call_args[0][0] == no_timeout_op
+
+    @pytest.mark.it("Sends ops that do need a timer to the next stage")
+    def test_sends_yes_timer_op_down(self, stage, mock_timer, yes_timeout_op):
+        stage.run_op(yes_timeout_op)
+        assert stage.next.run_op.call_count == 1
+        assert stage.next.run_op.call_args[0][0] == yes_timeout_op
+
+    @pytest.mark.it("Does not set a timer for ops that don't need a timer set")
+    def test_does_not_set_timer(self, stage, mock_timer, no_timeout_op):
+        stage.run_op(no_timeout_op)
+        assert mock_timer.call_count == 0
```

Comment:
Nit: for readability, I'd prefer this test to be directly after the related test (i.e. Sends ops that don't need a timer to the next stage")